lib/email/receiver.rb

```diff
@@ -660,102 +660,97 @@ def trim_discourse_markers(reply)
+        original_from_address, _ =
+          Email::Receiver.extract_email_address_and_name(mail[:x_original_from])
+        reply_to_address, reply_to_name =
+          Email::Receiver.extract_email_address_and_name(mail[:reply_to])
+        return reply_to_address, reply_to_name if original_from_address == reply_to_address
       end
 
-      return extract_from_address_and_name(mail.from) if mail.from.is_a? String
-
-      if mail.from.is_a? Mail::AddressContainer
-        mail.from.each do |from|
-          from_address, from_display_name = extract_from_address_and_name(from)
-          return from_address, from_display_name if from_address
-        end
-      end
-
-      nil
+      Email::Receiver.extract_email_address_and_name(mail[:from])
     rescue StandardError
       nil
     end
 
-    def extract_from_fields_from_header(mail_object, header, comparison_headers: [])
-      mail_object[header].each do |address_field|
-        from_address = address_field.address
-        from_display_name = address_field.display_name&.to_s
-
-        comparison_failed = false
-        comparison_headers.each do |comparison_header|
-          comparison_header_address = mail_object[comparison_header].to_s[/<([^>]+)>/, 1]
-          if comparison_header_address != from_address
-            comparison_failed = true
-            break
-          end
-        end
-
-        next if comparison_failed
-        next if !from_address&.include?("@")
-        return from_address&.downcase, from_display_name&.strip
-      end
+    def self.extract_email_address_and_name_from_mailman(mail)
+      list_address, _ = Email::Receiver.extract_email_address_and_name(mail[:list_post])
+      list_address, _ =
+        Email::Receiver.extract_email_address_and_name(mail[:x_beenthere]) if list_address.blank?
 
-      [nil, nil]
-    end
+      return if list_address.blank?
 
-    def extract_from_address_and_name(value)
-      if value[";"]
-        from_display_name, from_address = value.split(";")
-        return from_address&.strip&.downcase, from_display_name&.strip
-      end
+      # the CC header often includes the name of the sender
+      address_to_name = mail[:cc]&.element&.addresses&.to_h { [_1.address, _1.name] } || {}
 
-      if value[/<[^>]+>/]
-        from_address = value[/<([^>]+)>/, 1]
-        from_display_name = value[/\A([^<]+)/, 1]
+      %i[from reply_to x_mailfrom x_original_from].each do |header|
+        next if mail[header].blank?
+        email, name = Email::Receiver.extract_email_address_and_name(mail[header])
+        if email.present? && email != list_address
+          return email, name.presence || address_to_name[email]
+        end
       end
+    end
 
-      if (from_address.blank? || !from_address["@"]) && value[/\[mailto:[^\]]+\]/]
-        from_address = value[/\[mailto:([^\]]+)\]/, 1]
-        from_display_name = value[/\A([^\[]+)/, 1]
+    def self.extract_email_address_and_name(value)
```

Comment:
Refactored this method to use the battle-tested `Mail::Address` parser. I remember trying to do that in the past but it wasn't working every time. I found and handled all the cases that made the parser bail and also cleaned up surrounding quotes.